### PR TITLE
CI: Uffizzi Preview Environments for Testing Pull Requests

### LINKED ISSUE #3066: CI Feature Request: Integrate pull request preview environments
I would like to support Photoprism by implementing [Uffizzi](https://github.com/UffizziCloud/uffizzi) preview environments.
Disclaimer: I work on [Uffizzi](https://github.com/UffizziCloud/uffizzi).

Uffizzi is a Open Source full stack previews engine and our platform is available completely free for Photoprism (and all open source projects). This will provide maintainers with preview environments of every PR in the cloud, which enables faster iterations and reduces time to merge. You can see the open source repos which are currently using Uffizzi over [here](https://uffizzi.notion.site/)

Uffizzi is purpose-built for the task of previewing PRs and it integrates with your workflow to deploy preview environments in the background without any manual steps for maintainers or contributors.

We can go ahead and create an Initial PoC for you right away if you think there is value in this proposal.

### PR DESCRIPTION
**Pull request by ShrutiC-git**
This PR is adding support for preview environments for PR events (opened, sync'ed, reopened, etc) on Photoprism, powered by [Uffizzi](https://www.uffizzi.com/).

Fixes https://github.com/photoprism/photoprism/issues/3066

## Why This PR
`Photoprism` can greatly benefit from preview environments for PRs to fasten delivery speed and help the maintainers/contributors in the code-review process. 

Uffizzi is purpose-built for the task of previewing PRs and it integrates with existing workflow to deploy preview environments in the background without any manual steps for maintainers or contributors. It will add another layer of testing and reviewing on top of `Photoprism's` current testing stack, thus also cutting down on the time and effort it'd take to ensure proper coverage.

### What does the PR contain?

This PR uses two GHA workflows — `uffizzi-build.yml` and `uffizzi-preview.yml` — to build and provision preview environments. This PR uses `photoprism/bookworm/Dockerfile` to build `Photoprism` from source, ensuring changes are reflected in the preview environment. 

The `docker-compose.uffizzi.yml` defines `photoprism` container, `postgres` for storage, and also defines Uffizzi integration, which is an entry point into the `photoprism` container.

Once a PR event is triggered — PR opened, PR synced, review requested — the build workflow `(uffizzi-build.yml)` will build the containers defined in the `docker-compose.uffizzi.yml` file, and the preview workflow `(uffizzi-preview.yml)` will either spin up a new preview or update an existing preview. The previews are ephemeral — so they live for only as long as needed. A comment is posted on the PR containing the URL of the preview environment, allowing contributors and maintainers to easily navigate to the preview. 

Clicking the link posted on the comment will take the user to the preview. The preview comes will take users to the `sign-in` screen of `Photoprism`. The preview environment can be used like a prod instance of `Photoprism`.

### How was the PR Tested?

To test this, I opened a [PR against my fork](https://github.com/ShrutiC-git/photoprism/pull/9) of `Photoprism,` the `uffizzi-build.yml` file triggered the `Build PR Image` action. Once this action was completed, the `uffizzi-preview.yml` GHA-file triggered the `Deploy Preview` action. [This comment posted on the PR](https://github.com/ShrutiC-git/photoprism/pull/9#issuecomment-1402382555) takes me to the preview. This is the [preview environment](https://app.uffizzi.com/github.com/ShrutiC-git/photoprism/pull/9) for this PR. 

The preview environment took me to `Photoprism's` login screen. The environment uses the following credentials to sign-in:
```
username: admin
password: photoprism
```

Once logged in, I was able to use `Photoprism` like a production instance. This test preview environment is populated with test data to demonstrate the workings of Uffizzi preview environments. 

 PS: _We use this 2-step workflow for open-source projects, to allow open-source projects to have previews for contributions from forks. [Here](https://www.uffizzi.com/preview-environments-guide/github-actions-two-stage-workflow) is a more in-depth read into why we use a two-stage workflow._
 
If there is anything that needs to be changed on this PR or any other feedback/concerns, please do let me know! I believe Uffizzi will be a great addition to Photoprism. 

Acceptance Criteria:

- [X] **Features and enhancements are fully implemented** so that they can be released at any time without additional work
- [X] **Automated unit and/or acceptance tests have been added** to ensure the changes work as expected and to reduce repetitive manual work
- [ ] **User interface changes are fully responsive** and have been tested on all major browsers and various devices
- [ ] Database-related changes are compatible with SQLite and MariaDB
- [ ] Translations have been / will be updated (specify if needed)
- [ ] Documentation has been / will be updated (specify if needed)
- [X] Contributor License Agreement (CLA) has been signed

### CODE REVIEW

**Comment by graciousgrey** [comment]
Thank you very much for your efforts, we will look into this as soon as our time permits.

**Comment by ShrutiC-git** [comment]
thanks, @graciousgrey, sounds good. Let me know if anything needs more details or explanation!

**Comment by ShrutiC-git** [comment]
Hey @graciousgrey!

Just a friendly ping - would you be able to give a timeframe for looking at this? Curious to see what the community thinks of these preview environments. I'd love this integration to go through — I had such great fun developing this PoC and seeing it in action!
 
Let me know :)

File: docker-compose.uffizzi.yml [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+      # - "./pictures:/photoprism/originals"               # Original media files (DO NOT REMOVE)
+      # - "/example/family:/photoprism/originals/family" # *Additional* media folders can be mounted like this
+      # - "~/Import:/photoprism/import"                  # *Optional* base folder from which files can be imported to originals
+      - "storage:/photoprism/storage"                  # *Writable* storage folder for cache, database, and sidecar files (DO NOT REMOVE)
+    deploy:
+      resources:
+        limits:
+          memory: 2000M
+  postgres:
+    image: postgres:12-alpine
```
Comment (lastzero) [inline docker-compose.uffizzi.yml:103]: Unfortunately, only MariaDB and SQLite are supported at this time.

**Comment by lastzero** [comment]
@ShrutiC-git I apologize for the wait! We had to finish our work on some important features first, such as free high-resolution maps for all users. In the coming weeks we will focus on merging the open pull requests.

Looking at your solution, I have the following questions and comments:

- Would this only be used to test pull requests?
- If so, how is the service funded, e.g. does it require a paid account?
- Will it send emails to our contributors every time they commit a change? Some of them are very sensitive when it comes to privacy, so I imagine this could cause irritation.
- The [docker-compose.uffizzi.yml](https://github.com/photoprism/photoprism/pull/3144/files#diff-3f919f8a1945a404eff1ade023f33713229be85c92d9b39947b213748572d46b) file in your PR uses Postgres, but currently only MariaDB and SQLite are supported. So, at a minimum, that must be changed for a successful test.

**Comment by lastzero** [comment]
Unfortunately we haven't received any feedback on our questions and comments, so I'm closing this PR for now. Thank you very much for your contribution! :)

**Pull request closed without merging**